contracts/voting-with-uuid/src/contract.rs

```diff
@@ -0,0 +1,516 @@
+
+        let staked_weight = deps
+            .querier
+            .query_balance(&env.contract.address, &state.denom)
+            .unwrap()
+            .amount
+            .u128();
+
+        if staked_weight == 0 {
+            return Err(ContractError::PollNoStake {});
+        }
+
+        let quorum = ((tallied_weight / staked_weight) * 100) as u8;
+        if a_poll.quorum_percentage.is_some() && quorum < a_poll.quorum_percentage.unwrap() {
+            // Quorum: More than quorum_percentage of the total staked tokens at the end of the voting
+            // period need to have participated in the vote.
+            rejected_reason = "Quorum not reached";
+        } else if yes > tallied_weight / 2 {
+            //Threshold: More than 50% of the tokens that participated in the vote
+            // (after excluding “Abstain” votes) need to have voted in favor of the proposal (“Yes”).
+            a_poll.status = PollStatus::Passed;
+            passed = true;
+        } else {
+            rejected_reason = "Threshold not reached";
+        }
+    } else {
+        rejected_reason = "Quorum not reached";
+    }
+    if !passed {
+        a_poll.status = PollStatus::Rejected
+    }
+    poll(deps.storage).save(poll_id.as_slice(), &a_poll)?;
+
+    for voter in &a_poll.voters {
+        unlock_tokens(deps.storage, voter, poll_id)?;
+    }
+
+    let attributes = vec![
+        attr("action", "end_poll"),
+        attr("poll_id", poll_id.to_string()),
+        attr("rejected_reason", rejected_reason),
+        attr("passed", &passed),
+    ];
+
+    let r = Response {
+        submessages: vec![],
+        messages: vec![],
+        attributes,
+        data: None,
+    };
+    Ok(r)
+}
+
+// unlock voter's tokens in a given poll
+fn unlock_tokens(
+    storage: &mut dyn Storage,
+    voter: &Addr,
+    poll_id: Uuid,
+) -> Result<Response, ContractError> {
+    let voter_key = &voter.as_str().as_bytes();
+    let mut token_manager = bank_read(storage).load(voter_key).unwrap();
+
+    // unlock entails removing the mapped poll_id, retaining the rest
+    token_manager.locked_tokens.retain(|(k, _)| k != &poll_id);
+    bank(storage).save(voter_key, &token_manager)?;
+    Ok(Response::default())
+}
+
+// finds the largest locked amount in participated polls.
+fn locked_amount(voter: &[u8], storage: &dyn Storage) -> Uint128 {
+    let token_manager = bank_read(storage).load(voter).unwrap();
+    token_manager
+        .locked_tokens
+        .iter()
+        .map(|(_, v)| *v)
+        .max()
+        .unwrap_or_default()
+}
+
+fn has_voted(voter: &Addr, a_poll: &Poll) -> bool {
+    a_poll.voters.iter().any(|i| i == voter)
+}
+
+pub fn cast_vote(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    poll_id: Uuid,
+    vote: String,
+    weight: Uint128,
+) -> Result<Response, ContractError> {
+    let poll_key = poll_id.as_bytes();
+    let state = config_read(deps.storage).load()?;
```

Comment:
nit: unused value, remove it ?

Reply:
I left it in the changing to using uuid. Deleted. thanks!